backend/api_calls.py

```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
+    message: Message to send to client.
+    previewMessage: Short preview of the message conents. Optional.
+    app: Whether to send as a push notification to the iOS/Android/Web clients.
+    vibrate: Whether the notification should vibrate on iOS/Android.
+    sound: Whether the notification should play a sound on iOS/Android.
+    destination: URL that notification should open when clicked.
+    sendTime: When to spawn notifications from this template. UNIX timestamp.
+    playerId: Player id to send message to.
+    groupId: Who to send the notifications to.
+    icon: An icon code to show.
+
+  Firebase entries:
+    /notifications/%(notificationId)
+"""
+  valid_args = ['notificationId']
+  required_args = list(valid_args)
+  ValidateInputs(request, firebase, required_args, valid_args)
+
+  put_data = {}
+  properties = ['message', 'app', 'vibrate', 'sound', 'destination', 'sendTime',
+                'groupId', 'playerId', 'icon']
+
+  for property in properties:
+    if property in request:
+      put_data[property] = request[property]
+
+  return firebase.patch('/notifications/%s' % request['notificationId'], put_data)
+
+
+def MarkNotificationSeen(request, firebase):
+  """Updates the notification's seenTime.
+
+  Validation:
+    notificationId must exist.
+
+  Args:
+    notificationId: Notification Id to update.
+
+  Firebase entries:
+    /notifications/notifications/%(notificationId)
+"""
+  valid_args = ['notificationId', 'playerId']
+  required_args = list(valid_args)
+  ValidateInputs(request, firebase, required_args, valid_args)
+  put_data = {
+    'seenTime': time.time()
+  }
+  results = []
+  results.append(firebase.patch('/notifications/notifications/%s' % (
+      request['notificationId']), put_data))
+  return results
+
+
+def _handleNotification(self, firebase, notification_id, notification):
+  """Helper function to propogate a notification."""
+
+  # TODO: Send to ionic
+  if 'playerId' in notification:
+    firebase.put('/players/%s/notifications' % notification['playerId'],
+                  notification_id, notification)
+  elif 'groupId' in notification:
+    players = firebase.get('/groups/%s' % notification['groupId'],
+                           'players')
+    logging.info(players)
+    for player in players:
+      firebase.put('/players/%s/notifications' % player,
+                   notification_id, notification)
+  else:
+    logging.error('Notification does not have a playerId or a groupId!')
+
+
+def ExecuteNotifications(request, firebase):
+  """INTERNAL ONLY: Send the notifications.
+
+  This method will execute queued notifications and send them to apps and add
+  them to the player specific notification database URLs.
+  """
+  # Handle notifications, 20 at a time so we don't need to get the entire
+  # notification database.
+  get_params = {
+      #
+      'orderBy': '"sent"',
+      'limitToLast': 20
+  }
+  updates = True
+  while updates:
```

Comment:
It keeps checking until it doesn't send any more notifications. Modified code a bit to make this more clear.